internal/querynode/stage_unsolved.go

```diff
@@ -0,0 +1,476 @@
+// Copyright (C) 2019-2020 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License.
+
+package querynode
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	oplog "github.com/opentracing/opentracing-go/log"
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/msgstream"
+	"github.com/milvus-io/milvus/internal/proto/commonpb"
+	"github.com/milvus-io/milvus/internal/proto/segcorepb"
+	"github.com/milvus-io/milvus/internal/util/timerecord"
+	"github.com/milvus-io/milvus/internal/util/trace"
+	"github.com/milvus-io/milvus/internal/util/tsoutil"
+)
+
+type unsolvedStage struct {
+	ctx context.Context
+
+	collectionID UniqueID
+	vChannel     Channel
+
+	input  chan queryMsg
+	output chan queryResult
+
+	streaming         *streaming
+	queryResultStream msgstream.MsgStream
+
+	unsolvedMsgMu sync.Mutex // guards unsolvedMsg
+	unsolvedMsg   []queryMsg
+
+	tSafeWatcher *tSafeWatcher
+
+	serviceableTimeMutex sync.Mutex // guards serviceableTime
+	serviceableTime      Timestamp
+}
+
+func newUnsolvedStage(ctx context.Context,
+	collectionID UniqueID,
+	vChannel Channel,
+	input chan queryMsg,
+	output chan queryResult,
+	streaming *streaming,
+	queryResultStream msgstream.MsgStream) *unsolvedStage {
+
+	return &unsolvedStage{
+		ctx:               ctx,
+		collectionID:      collectionID,
+		vChannel:          vChannel,
+		input:             input,
+		output:            output,
+		unsolvedMsg:       make([]queryMsg, 0),
+		streaming:         streaming,
+		queryResultStream: queryResultStream,
+	}
+}
+
+func (q *unsolvedStage) start() {
+	q.register()
+	go q.doUnsolvedQueryMsg()
+	for {
+		select {
+		case <-q.ctx.Done():
+			log.Debug("stop unsolvedStage", zap.Int64("collectionID", q.collectionID))
+			return
+		case msg := <-q.input:
+			msgType := msg.Type()
```

Comment:
This step is redundant and you can put the message directly into unsolved channel, which is processed by do unsolvedQueryMsg()

Reply:
fixed